File: src/ios/JAnalyticsPlugin.m
#import "JAnalyticsPlugin.h"
#import <CoreLocation/CoreLocation.h>

#import "JANALYTICSService.h"
#import "JANALYTICSEventObject.h"

// 如果需要使用idfa功能所需要引入的头文件（可选）
//#import <AdSupport/AdSupport.h>

@implementation JAnalyticsPlugin

- (void)init:(CDVInvokedUrlCommand *)command {
    NSDictionary *plistDic = [[NSBundle mainBundle] infoDictionary];
    NSString* APP_KEY = [[plistDic objectForKey:@"JAnalytics"] objectForKey:@"APP_KEY"];
    NSString* CHANNEL = [[plistDic objectForKey:@"JAnalytics"] objectForKey:@"CHANNEL"];
    
    JANALYTICSLaunchConfig * config = [[JANALYTICSLaunchConfig alloc] init];
    config.appKey = APP_KEY;
    config.channel = CHANNEL;
    [JANALYTICSService setupWithConfig:config];
}

- (void)setDebugMode:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    bool enable = params[@"enable"];
    [JANALYTICSService setDebug:enable];
}

- (void)initCrashHandler:(CDVInvokedUrlCommand *)command {
    [JANALYTICSService crashLogON];
}

- (void)stopCrashHandler:(CDVInvokedUrlCommand *)command {
    
}

- (void)onPageStart:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    NSString *pageName = params[@"pageName"];
    [JANALYTICSService startLogPageView:pageName];
}

- (void)onPageEnd:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    NSString *pageName = params[@"pageName"];
    [JANALYTICSService stopLogPageView:pageName];
}

- (void)addCountEvent:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    NSString *eventId = params[@"eventId"];
    NSDictionary<NSString *, NSString *> *extra = params[@"extras"];
    
    JANALYTICSCountEvent *event = [[JANALYTICSCountEvent alloc] init];
    event.eventID = eventId;
    event.extra = extra;
    [JANALYTICSService eventRecord:event];
}

- (void)addCalculateEvent:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    NSString *eventId = params[@"eventId"];
    NSNumber *eventValue = params[@"eventValue"];
    NSDictionary<NSString *, NSString *> *extra = params[@"extras"];
    
    JANALYTICSCalculateEvent *event = [[JANALYTICSCalculateEvent alloc] init];
    event.eventID = eventId;
    event.value = [eventValue floatValue];
    event.extra = extra;
    [JANALYTICSService eventRecord:event];
}

- (void)addLoginEvent:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    NSString *loginMethod = params[@"loginMethod"];
    NSNumber *isLoginSuccess = params[@"isLoginSuccess"];
    NSDictionary<NSString *, NSString *> *extra = params[@"extras"];
    
    JANALYTICSLoginEvent *event = [[JANALYTICSLoginEvent alloc] init];
    event.method = loginMethod;
    event.success = isLoginSuccess;
    event.extra = extra;
    [JANALYTICSService eventRecord:event];
}

- (void)addRegisterEvent:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    NSString *registerMethod = params[@"registerMethod"];
    NSNumber *isRegisterSuccess = params[@"isRegisterSuccess"];
    NSDictionary<NSString *, NSString *> *extra = params[@"extras"];
    
    JANALYTICSRegisterEvent *event = [[JANALYTICSRegisterEvent alloc] init];
    event.method = registerMethod;
    event.success = isRegisterSuccess;
    event.extra = extra;
    [JANALYTICSService eventRecord:event];
}

- (void)addBrowseEvent:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    
    NSString *browseId = params[@"browseId"];
    NSString *browseName = params[@"browseName"];
    NSString *browseType = params[@"browseType"];
    NSNumber *browseDuration = params[@"browseDuration"];
    NSDictionary<NSString *, NSString *> *extra = params[@"extras"];
    
    JANALYTICSBrowseEvent * event = [[JANALYTICSBrowseEvent alloc] init];
    event.contentID = browseId;
    event.name = browseName;
    event.type = browseType;
    event.duration = [browseDuration longValue];
    event.extra = extra;
    [JANALYTICSService eventRecord:event];
}

- (void)addPurchaseEvent:(CDVInvokedUrlCommand *)command {
    NSDictionary *params = [command.arguments objectAtIndex:0];
    
    NSString *goodsId = params[@"goodsId"];
    NSString *goodsName = params[@"goodsName"];
    NSNumber *price = params[@"price"];
    NSString *currency = params[@"currency"];
    NSNumber *isPurchaseSuccess = params[@"isPurchaseSuccess"];
    NSString *goodsType = params[@"goodsType"];
    NSNumber *goodsCount = params[@"goodsCount"];
    NSDictionary<NSString *, NSString *> *extra = params[@"extras"];
    
    JANALYTICSPurchaseEvent * event = [[JANALYTICSPurchaseEvent alloc] init];
    event.success = isPurchaseSuccess;
    event.price = [price floatValue];
    event.goodsName = goodsName;
    event.goodsType = goodsType;
    event.quantity = [goodsCount integerValue];
    event.goodsID = goodsId;
    event.currency = [currency isEqual:@"USD"] ? JANALYTICSCurrencyUSD: JANALYTICSCurrencyCNY;
    event.extra = extra;
    [JANALYTICSService eventRecord:event];
}

- (void)setAnalyticsReportPeriod:(CDVInvokedUrlCommand *)command {
  NSDictionary *params = [command.arguments objectAtIndex:0];
  [JANALYTICSService setFrequency:[params[@"period"] unsignedIntegerValue]];
}

- (void)identifyAccount:(CDVInvokedUrlCommand *)command {
  NSDictionary *params = [command.arguments objectAtIndex:0];
  
  JANALYTICSUserInfo *userInfo = [[JANALYTICSUserInfo alloc] init];
  userInfo.accountID = params[@"accountID"];
  userInfo.name = params[@"name"];
  userInfo.creationTime = [params[@"creationTime"] doubleValue];
  if ([params[@"sex"] intValue] == 1) {
    userInfo.sex = JANALYTICSSexMale;
  } else if ([params[@"creationTime"] intValue] == 2) {
    userInfo.sex = JANALYTICSSexFemale;
  } else {
    userInfo.sex = JANALYTICSSexUnknown;
  }
  
  if ([params[@"paid"] intValue] == 1) {
    userInfo.paid = JANALYTICSPaidPaid;
  } else if ([params[@"creationTime"] intValue] == 2) {
    userInfo.paid = JANALYTICSPaidUnpaid;
  } else {
    userInfo.paid = JANALYTICSPaidUnknown;
  }
  
  userInfo.birthdate = params[@"birthdate"];
  userInfo.phone = params[@"phone"];
  userInfo.email = params[@"email"];
  userInfo.weiboID = params[@"weiboID"];
  userInfo.wechatID = params[@"wechatID"];
  userInfo.qqID = params[@"qqID"];
  NSDictionary *extras = params[@"extras"];
  for (NSString* key in extras.allKeys) {
    [userInfo setExtraObject:extras[key] forKey:key];
  }
  
  [JANALYTICSService identifyAccount:userInfo with:^(NSInteger err, NSString *msg) {
    CDVPluginResult *result;
    if (err) {
       result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString: msg];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    } else {

      result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
  }];
}

@end

File: src/android/JAnalyticsPlugin.java
package cn.jiguang.cordova.analytics;

import android.content.Context;
import android.util.Log;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import cn.jiguang.analytics.android.api.Account;
import cn.jiguang.analytics.android.api.AccountCallback;
import cn.jiguang.analytics.android.api.BrowseEvent;
import cn.jiguang.analytics.android.api.CalculateEvent;
import cn.jiguang.analytics.android.api.CountEvent;
import cn.jiguang.analytics.android.api.Currency;
import cn.jiguang.analytics.android.api.JAnalyticsInterface;
import cn.jiguang.analytics.android.api.LoginEvent;
import cn.jiguang.analytics.android.api.PurchaseEvent;
import cn.jiguang.analytics.android.api.RegisterEvent;

public class JAnalyticsPlugin extends CordovaPlugin {

    private ExecutorService threadPool = Executors.newFixedThreadPool(1);

    private Context mContext;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        mContext = cordova.getActivity().getApplicationContext();
    }

    @Override
    public boolean execute(final String action, final JSONArray data, final CallbackContext callback)
            throws JSONException {
        threadPool.execute(new Runnable() {

            @Override
            public void run() {
                try {
                    Method method = JAnalyticsPlugin.class.getDeclaredMethod(action, JSONArray.class,
                            CallbackContext.class);
                    method.invoke(JAnalyticsPlugin.this, data, callback);
                } catch (NoSuchMethodException e) {
                    e.printStackTrace();
                } catch (InvocationTargetException e) {
                    e.printStackTrace();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                }
            }
        });

        return true;
    }

    void init(JSONArray data, CallbackContext callback) {
        JAnalyticsInterface.init(mContext);
    }

    void setDebugMode(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        boolean enable = params.getBoolean("enable");
        JAnalyticsInterface.setDebugMode(enable);
    }

    void initCrashHandler(JSONArray data, CallbackContext callback) {
        JAnalyticsInterface.initCrashHandler(mContext);
    }

    void stopCrashHandler(JSONArray data, CallbackContext callback) {
        JAnalyticsInterface.stopCrashHandler(mContext);
    }

    void setChannel(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String channel = params.getString("channel");
        JAnalyticsInterface.setChannel(mContext, channel);
    }

    void onPageStart(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String pageName = params.getString("pageName");
        JAnalyticsInterface.onPageStart(mContext, pageName);
    }

    void onPageEnd(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String pageName = params.getString("pageName");
        JAnalyticsInterface.onPageEnd(mContext, pageName);
    }

    void addCountEvent(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String eventId = params.getString("eventId");
        JSONObject extras = params.has("extras") ? params.getJSONObject("extras") : null;

        CountEvent event = new CountEvent(eventId);
        if (extras != null) {
            event.addExtMap(toMap(extras));
        }
        JAnalyticsInterface.onEvent(mContext, event);
    }

    void addCalculateEvent(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String eventId = params.getString("eventId");
        double eventValue = params.getDouble("eventValue");
        JSONObject extras = params.has("extras") ? params.getJSONObject("extras") : null;

        CalculateEvent event = new CalculateEvent(eventId, eventValue);
        if (extras != null) {
            event.addExtMap(toMap(extras));
        }
        JAnalyticsInterface.onEvent(mContext, event);
    }

    void addLoginEvent(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String loginMethod = params.getString("loginMethod");
        boolean isLoginSuccess = params.getBoolean("isLoginSuccess");
        JSONObject extras = params.has("extras") ? params.getJSONObject("extras") : null;

        LoginEvent event = new LoginEvent(loginMethod, isLoginSuccess);
        if (extras != null) {
            event.addExtMap(toMap(extras));
        }
        JAnalyticsInterface.onEvent(mContext, event);
    }

    void addRegisterEvent(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String registerMethod = params.getString("registerMethod");
        boolean isRegisterSuccess = params.getBoolean("isRegisterSuccess");
        JSONObject extras = params.has("extras") ? params.getJSONObject("extras") : null;

        RegisterEvent event = new RegisterEvent(registerMethod, isRegisterSuccess);
        if (extras != null) {
            event.addExtMap(toMap(extras));
        }
        JAnalyticsInterface.onEvent(mContext, event);
    }

    void addBrowseEvent(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String browseId = params.getString("browseId");
        String browseName = params.getString("browseName");
        String browseType = params.getString("browseType");
        long browseDuration = params.getLong("browseDuration");
        JSONObject extras = params.has("extras") ? params.getJSONObject("extras") : null;

        BrowseEvent event = new BrowseEvent(browseId, browseName, browseType, browseDuration);
        if (extras != null) {
            event.addExtMap(toMap(extras));
        }
        JAnalyticsInterface.onEvent(mContext, event);
    }

    void addPurchaseEvent(JSONArray data, CallbackContext callback) throws JSONException {
        JSONObject params = data.getJSONObject(0);
        String goodsId = params.getString("goodsId");
        String goodsName = params.getString("goodsName");
        double price = params.getDouble("price");
        boolean isPurchaseSuccess = params.getBoolean("isPurchaseSuccess");
        String goodsType = params.getString("goodsType");
        int goodsCount = params.getInt("goodsCount");

        String currencyStr = params.getString("currency");
        Currency currency = Currency.CNY;
        if (currencyStr.equals("USD")) {
            currency = Currency.USD;
        }

        JSONObject extras = params.has("extras") ? params.getJSONObject("extras") : null;

        PurchaseEvent event = new PurchaseEvent(goodsId, goodsName, price, isPurchaseSuccess, currency, goodsType,
                goodsCount);
        if (extras != null) {
            event.addExtMap(toMap(extras));
        }

        JAnalyticsInterface.onEvent(mContext, event);
    }

    void setAnalyticsReportPeriod(JSONArray data, CallbackContext callback) throws JSONException {// ......
        JSONObject params = data.getJSONObject(0);
        int period = params.getInt("period");
        JAnalyticsInterface.setAnalyticsReportPeriod(mContext, period);
    }

    void identifyAccount(JSONArray data, CallbackContext callback) throws JSONException {// ......
        JSONObject params = data.getJSONObject(0);
        String accountID = params.getString("accountID");
        long creationTime = params.getLong("creationTime");
        String name = params.getString("name");
        int sex = params.getInt("sex");
        int paid = params.getInt("paid");
        String birthdate = params.getString("birthdate");
        String phone = params.getString("phone");
        String email = params.getString("email");
        String weiboID = params.getString("weiboID");
        String wechatID = params.getString("wechatID");
        String qqID = params.getString("qqID");
        JSONObject extras = params.has("extras") ? params.getJSONObject("extras") : null;

        Account account = new Account(accountID);
        account.setCreationTime(creationTime); // 账户创建的时间戳
        account.setName(name);
        account.setSex(sex);
        account.setPaid(paid);
        account.setBirthdate(birthdate); // "19880920"是yyyyMMdd格式的字符串
        account.setPhone(phone);
        account.setEmail(email);
        account.setWeiboId(weiboID);
        account.setWechatId(wechatID);
        account.setQqId(qqID);
        if (extras != null) {
            Iterator iterator = extras.keys();
            while (iterator.hasNext()) {
                String key = (String) iterator.next();
                String value = extras.getString(key);
                account.setExtraAttr(key, value); // key如果为空，或者以极光内部namespace(符号$)开头，会设置失败并打印日志
            }
        }
        JAnalyticsInterface.identifyAccount(mContext, account, new AccountCallback() {
            @Override
            public void callback(int code, String msg) {
                if (code == 0) {
                    callback.success();
                } else {
                    Log.i(JAnalyticsPlugin.class.getSimpleName(), "Identify account error code " + code + " :" + msg);
                    callback.error(msg);
                }
            }
        });
    }

    private Map<String, String> toMap(JSONObject json) throws JSONException {
        Map<String, String> map = new HashMap<String, String>();
        Iterator iterator = json.keys();
        while (iterator.hasNext()) {
            String key = (String) iterator.next();
            String value = json.getString(key);
            map.put(key, value);
        }
        return map;
    }
}
